File: descripe/imapp_modisl2/src/leocat/src/MOD_PRAlg29/chk_coast.c
/*
************************************************************************
!F77 

!Description:

   Routine which checks for extreme values of NDVI in coastal situations.
   Sets confidence to "confident clear" if very low or
   very high values are found (clear water or clear land, respectively).

!Input parameters:
   none

   Inputs stored in structure variable 'pxin' of type "pixel_in" defined
   in pixel.h. 

!Output Parameters:
   conf        modified clear sky confidence for pixel

   Returned through function call

!Revision History:
    Converted to C         11/07      R. Frey

!Team-unique Header:

!References and Credits:
   See Cloud Mask ATBD.

!END
***********************************************************************/

/* Includes */

#include <stdio.h>
#include <math.h>
#include "pixel.h"
#include "thresholds.h"
#include "mask_processing_constants.h"


float chk_coast()


{

/***********************************************************************/
	
//  Declarations
	
    extern int check_bits(int, unsigned char[]);
    extern void set_bit(int, unsigned char[]);
    
    float m01, m02;
    float ndvi;
    float conf;
    
    int bit_test;
    int qa_test;
    int irclr;
    
/**********************************************************************/

//  Initializations

//  printf("Executing chk_coast \n");
    conf = pxout.intermediate_conf;
    
    ndvi = 0.0;
    
    m01 = pxin.rad[0];
    m02 = pxin.rad[1];
    
/**********************************************************************/    
    
//  Determine the logical flag 'irclr' - true if ir cloud tests below
//  have all been passed. IR thin cirrus test makes final decision for
//  bit 18.
    
    irclr = 1;
    bit_test = check_bits(14, pxout.testbits);
    qa_test = check_bits(14, pxout.qabits);
    if( qa_test && !bit_test ) irclr = 0;
    bit_test = check_bits(15, pxout.testbits);
    qa_test = check_bits(15, pxout.qabits);
    if( qa_test && !bit_test ) irclr = 0;
    bit_test = check_bits(18, pxout.testbits);
    qa_test = check_bits(18, pxout.qabits);
    if( qa_test && !bit_test ) irclr = 0;

    if(irclr) {
      if ( rintf(m02) != rintf(bad_data) &&
              rintf(m01) != rintf(bad_data) ) {

//      Check for very low or very high ndvi values.
        (void) set_bit(22, pxout.qabits);
        ndvi = (m02 - m01) / (m02 + m01);
        if(ndvi <= swc_ndvi[0] || ndvi >= swc_ndvi[1]) {
          conf = 1.0;
          (void) set_bit(22, pxout.testbits);
        }

      }
    }

/**********************************************************************/

//  printf("CC NDVI: %d %f %f %f %f %f %f \n", irclr, m01, m02, ndvi,
//  		swc_ndvi[0], swc_ndvi[1], conf);

    return conf;
    
/**********************************************************************/    
    
}

File: descripe/imapp_modisl2/src/cloudmask/src/chk_coast.f
      subroutine chk_coast(pxldat,confdnc,qa_bits,testbits)


c---------------------------------------------------------------------
C!F77 
c
c!Description:
c
c     Routine which checks for extreme values of NDVI in coastal 
c     regions.  Set confidence to "confident clear" if very low or
c     very high values are found.
c
c!Input parameters:
c pxldat        Array containing reflectance or brightness temperatures
c               for all bands for current pixel
c confdnc       Current pixel unobstructed confidence
c
c!Output Parameters:
c qa_bits       Byte array containing qa bits
c testbits      Byte array containing test results
c
c!Revision History:
c
c!Team-unique Header:
c
c!References and Credits:
c See Cloud Mask ATBD-MOD-35.
c
c!END
c---------------------------------------------------------------------

      save 

      include 'global.inc'
      include 'swc_ndvi.inc'

c     Scalar arguments
      real confdnc

c     Array arguments
      real pxldat(inband)
      byte testbits(6),qa_bits(10)

c     Local scalars
      integer rtn,debug,h_output
      real ndvi
      logical irclr

c     External subroutines
      external check_bits,set_bit,set_qa_bit

c     Common statement for debug purposes
      common / bug / debug, h_output

c---------------------------------------------------------------------

c     Determine the logical flag 'irclr' - true if ir cloud tests below
c     have all been passed. APOLLO test makes final decision for bit 18.

      irclr = .true.
      call check_bits(testbits,14,rtn)
      if(rtn .eq. 0) irclr = .false.
      call check_bits(testbits,15,rtn)
      if(rtn .eq. 0) irclr = .false.
      call check_bits(testbits,18,rtn)
      if(rtn .eq. 0) irclr = .false.

      if(irclr) then
        if (nint(pxldat(2)) .ne. nint(bad_data) .and.
     +           nint(pxldat(1)) .ne. nint(bad_data)) then

c         Check for very low or very high ndvi values.
          call set_qa_bit(qa_bits,22)
          ndvi = (pxldat(2) - pxldat(1)) / (pxldat(2) + pxldat(1))
          if(ndvi .le. swc_ndvi(1) .or. ndvi .ge. swc_ndvi(2)) then
            confdnc = 1.0
            call set_bit(testbits,22)
          end if

        end if
      end if

c---------------------------------------------------------------------

c     Debug statement.
      if(debug .gt. 0) then
        write(h_output,'(10x,'' Coastal NDVI test results: '')')
        write(h_output,'(10x,'' NDVI thresholds: '',2f10.5)') swc_ndvi
        write(h_output,'(10x,''irclr,ndvi,confdnc: '',l5,2f10.5/)')
     *         irclr,ndvi,confdnc
      end if

c---------------------------------------------------------------------

      return
      end
